cni/pkg/nftables/host_sets.go

```diff
@@ -0,0 +1,385 @@
+
+	// Flush IPv4 set
+	tx.Flush(&knftables.Set{
+		Name: h.v4SetName,
+	})
+
+	// Flush IPv6 set if enabled
+	if h.enableIPv6 {
+		tx.Flush(&knftables.Set{
+			Name: h.v6SetName,
+		})
+	}
+
+	if err := nft.Run(context.TODO(), tx); err != nil {
+		return fmt.Errorf("failed to flush nftables sets: %w", err)
+	}
+
+	builder.LogNftRules(tx)
+	return nil
+}
+
+// DestroySet completely destroys both sets and the table
+func (h *HostNftSetManager) DestroySet() error {
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return err
+	}
+
+	tx := nft.NewTransaction()
+
+	// Delete the entire table (which will delete sets and rules)
+	tx.Delete(&knftables.Table{})
+
+	if err := nft.Run(context.TODO(), tx); err != nil {
+		return fmt.Errorf("failed to destroy nftables table %s: %w", AmbientNatTable, err)
+	}
+
+	builder.LogNftRules(tx)
+	return nil
+}
+
+// ClearEntriesWithComment removes all entries with the specified comment
+func (h *HostNftSetManager) ClearEntriesWithComment(comment string) error {
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return err
+	}
+
+	// A pod can have multiple IPs, so we have to clear it from both v4 and v6 sets.
+	if err := h.clearEntriesFromSetWithComment(nft, h.v4SetName, comment); err != nil {
+		return fmt.Errorf("failed to clear entries from IPv4 address set: %w", err)
+	}
+
+	// Clear from IPv6 set if enabled
+	if h.enableIPv6 {
+		if err := h.clearEntriesFromSetWithComment(nft, h.v6SetName, comment); err != nil {
+			return fmt.Errorf("failed to clear entries from IPv6 address set: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ClearEntriesWithIP removes all entries with the specified IP address
+func (h *HostNftSetManager) ClearEntriesWithIP(ip netip.Addr) error {
+	return h.DeleteIP(ip, 0)
+}
+
+// clearEntriesWithIPAndComment takes an IP and a comment string and deletes any entries where both match the entry.
+//
+// Returns a non-nil error if listing or deletion fails.
+// For the first matching IP found in the list, *only* removes the entry if *both* the IP and comment match.
+// If the IP matches but the comment does not, returns the actual comment found to the caller, and does not remove any entries.
+// Otherwise, returns an empty string.
+func (h *HostNftSetManager) ClearEntriesWithIPAndComment(ip netip.Addr, comment string) (string, error) {
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return "", err
+	}
+
+	ipToCheck := ip.Unmap()
+	setName := h.v4SetName
+	if ipToCheck.Is6() {
+		if !h.enableIPv6 {
+			return "", fmt.Errorf("request to delete %s from the set, but ipv6 is not enabled", ipToCheck.String())
+		}
+		setName = h.v6SetName
+	}
+
+	elements, err := nft.ListElements(context.TODO(), "set", setName)
+	if err != nil {
+		return "", fmt.Errorf("failed to list elements from set %s: %w", setName, err)
+	}
+
+	// Find the element with matching IP
+	for _, elem := range elements {
```

Comment:
We have to loop through every IP in the set just to delete it? Ideally, using nftables would prevent us from doing these O(N) lookups.

Reply:
Ah this is great context thank you! Didn't realize we put the uid in comments